ct/l0: Introduce accurate per-stage write pipeline accounting

wdberkeley opened the pull request:
Previously, the write pipeline maintained a _current_sz member that counted the size of all write requests in the pipeline. It was incremented just before the write was added to the pipeline and decremented when request finished and its owning write_and_debounce call returned. The big problem with _current_sz is that it was being used for the data threshold, despite counting, e.g., bytes being uploaded, instead of just bytes waiting to be batched in the scheduler stage.

This change introduces careful accounting of write request bytes by stage, and uses this accurate accounting in the data threshold. A write request belongs to the pipeline whose pending request list it is linked into, and to the stage in its stage member. This means, for example, that a write request that is being proxied belongs to no pipeline and no stage, and isn't counted. The proxy request created from it does count. This differs from _current_sz accounting.

Note that the bytes of write request a shard holds in memory is not the same thing as the bytes owned by the pipeline, because proxied requests are still held and accounted for in memory on the source shard, but not on the target shard. It's the opposite for pipeline stage byte accounting. It's OK not to count the proxied request because the batch payload stays in memory on the source shard.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [X] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

* none

wdberkeley committed:
ct/l0: Introduce accurate per-stage write pipeline accounting

Previously, the write pipeline maintained a _current_sz member that
counted the size of all write requests in the pipeline. It was
incremented just before the write was added to the pipeline and
decremented when request finished and its owning write_and_debounce call
returned. The big problem with _current_sz is that it was being used for
the data threshold, despite counting, e.g., bytes being uploaded,
instead of just bytes waiting to be batched in the scheduler stage.

This change introduces careful accounting of write request bytes by
stage, and uses this accurate accounting in the data threshold. A write
request belongs to the pipeline whose pending request list it is linked
into, and to the stage in its stage member. This means, for example,
that a write request that is being proxied belongs to no pipeline and no
stage, and isn't counted. The proxy request created from it does count.
This differs from _current_sz accounting.

Note that the bytes of write request a shard holds in memory is not the
same thing as the bytes owned by the pipeline, because proxied requests
are still held and accounted for in memory on the source shard, but not
on the target shard. It's the opposite for pipeline stage byte
accounting. It's OK not to count the proxied request because the batch
payload stays in memory on the source shard.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--bf35a72d-cdcf-4d5a-bb95-98b3ea6ce193-->
<!--TEST RESULTS-->
<!--4cb70c0a-6e22-4f1a-b0a6-90afaa3d1f4d-->
<!--TEST RESULTS-->
<!--01e86e07-87e2-420f-a203-06093329e819-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77792">build#77792</a></summary>


| test_class                                                                                       | test_method                                                                                      | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:-------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_rpbench_test | src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_rpbench_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77792#019b13bb-6cd1-4842-b657-83a57efb4804 | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77924">build#77924</a></summary>


| test_class          | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:--------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77924#019b258e-b78c-455c-b043-772ed16214d9 | FLAKY         | 1/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.9265, p0=0.8357, reject_threshold=0.0100. adj_baseline=0.9996, p1=0.0040, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77924#019b2591-d6a8-47fd-bb2d-c7d0010d2b83 | FAIL          | 0/51     | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.9265, p0=0.0220, reject_threshold=0.0100. adj_baseline=0.9996, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77996">build#77996</a></summary>


| test_class                          | test_method             | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                     |
|:------------------------------------|:------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ScalingUpTest                       | test_fast_node_addition | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/77996#019b2948-f45f-42c7-bb63-78e74a921245 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0369, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1067, p1=0.3235, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition              |
| WriteCachingFailureInjectionE2ETest | test_crash_all          | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77996#019b2948-f462-4282-9a62-5c0ad05bdb18 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0765, p0=0.5487, reject_threshold=0.0100. adj_baseline=0.2123, p1=0.3397, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

Lazin commented:
The `write_request_scheduler_rpbench` fails with 

```
ERROR 2025-12-12 18:07:33,619 [shard 0:main] assert - Assert failure: (src/v/cloud_topics/level_zero/pipeline/write_pipeline.cc:396) '_stage_bytes[i] == computed[i]' stage_bytes drift detected: stage 0 has 1226 bytes recorded but 0 bytes computed from requests
```

Lazin commented on src/v/cloud_topics/level_zero/pipeline/write_pipeline.cc:
```diff
@@ -318,6 +341,63 @@ write_pipeline<Clock>::stage::choose_abort_source(ss::abort_source* maybe_as) {
+    }
+    if (to != unassigned_pipeline_stage) {
+        _stage_bytes[static_cast<size_t>(to()->get_numeric_id())] += bytes;
+    }
```
maybe it makes sense to add a dedicated counter for unassigned_pipeline_stage? Right now if we're transferring to unassinged the information is lost.

Lazin commented on src/v/cloud_topics/level_zero/pipeline/write_pipeline.cc:
```diff
@@ -231,6 +245,15 @@ void write_pipeline<Clock>::stage::push_next_stage(
 }
 
+template<class Clock>
+void write_pipeline<Clock>::stage::proxy_request(write_request<Clock>& req) {
```
Looking at the description

```
        /// Introduce a write request to the pipeline that is a proxy of a
        /// request from another pipeline. 
```

maybe it makes sense to call the method differently.  To me the name `proxy_request` reads like the request which is passed as a parameter will be copied, the proxy request will be created and then passed down the pipeline. But the method assumes that the proxy request already exists and is passed as a parameter. So maybe it should be called 'enqueue_proxy_request`?

```
        /// The request should not be linked and
        /// its stage should be unassigned. 
```

That could probably be an assertion in the code.


```
        The request will enter the pipeline
        /// at this stage's next stage.
```
The `write_pipeline::stage` has `push_next_stage` method which calls `reenqueue` method of the pipeline. It's not enough to add a request to the pending list. The condition variable should also be signaled to wake up the next stage.

wdberkeley replied:
Good suggestions on the name and assertions. The CV is specifically not signaled because the stage is signaled once after all the proxied requests have been enqueued (this is preserved from the original code). I can document that expectation in the comment.

Lazin commented on src/v/cloud_topics/level_zero/pipeline/write_pipeline.h:
```diff
@@ -209,8 +237,12 @@ class write_pipeline
-    // Current bytes (gauge)
-    size_t _current_size{0};
+    // Bytes per pipeline stage.
+    std::array<size_t, max_pipeline_stages> _stage_bytes{};
```
Maybe it makes sense to embed counters into stages or into pipeline_stage_container?

Lazin commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -350,10 +353,9 @@ write_request_scheduler<Clock>::proxy_write_request(
+      req->expiration_time);
     auto fut = proxy.response.get_future();
-    _stage.push_next_stage(proxy, false);
+    _stage.proxy_request(proxy);
```
I think that the accounting should be done inside the push_next_stage instead of doing it in a separate method. Currently, the definition of the next stage is to be relative to the request itself but it could be relative to the stage itself. This will allow to add request with 'unassigned' stage.

I don't fully understand how that works and what is the idea. So far my understanding is:
- if the request is added to any stage we should add its bytes to that stage's counter;
- if the request is removed from any stage we should take from the counter;
- unassigned is the exception, the code is allowed to add the request with any byte count if it was previously unassigned;

Following this logic the `push_next_stage` could just put the request to the pending queue and assign the stage to be the one that follows the current one and increment the corresponding counter.

wdberkeley replied:
If requests entered in the first stage, moved from one stage to the next in a single pipeline until the end of the pipeline, and then were retired, then the accounting could be simpler like this. The edges you're seeing are there because
1. some requests time out and get ejected from the pipeline early
2. some requests get pulled from their current pipeline and inserted into another at a later stage (note that the original code inserted proxied requests at the scheduler stage and advanced them to the next stage, but doing this can erroneously pop the data threshold because those proxied requests will count towards it in the scheduler stage even though they are about to move on to being batched).

`enqueue_proxy_request` could be done with `push_next_stage`... the different code is meant stylistically to handle a special case in a special way. But the timeout cleanup is trickier to handle cleanly, especially when remove_requests is implemented in the base_pipeline.

Lazin replied:
Maybe it makes sense to separate memory accounting from the accounting of work?
With per-stage accounting we're accounting for work that has to be done. The memory usage may not match exactly in this case. For instance, we're proxying one write request that has 1MiB of data to another shard. That shard doesn't receive 1MiB of data. The data is still allocated on the original shard. But instead it receives 1MiB of units of work (bytes to be uploaded). Maybe if these two things are cleanly separated it'd be more readable?

wdberkeley commented:
> The `write_request_scheduler_rpbench` fails with
> 
> ```
> ERROR 2025-12-12 18:07:33,619 [shard 0:main] assert - Assert failure: (src/v/cloud_topics/level_zero/pipeline/write_pipeline.cc:396) '_stage_bytes[i] == computed[i]' stage_bytes drift detected: stage 0 has 1226 bytes recorded but 0 bytes computed from requests
> ```

The double-check on the accounting is actually kind of obviously broken. If we time out requests, they get removed from the pending list but their stage bytes aren't subtracted until the defer in write_and_debounce runs. The request's write_and_debounce coroutine might not get scheduled for a bit after its future is resolved by the time out code, and in that period we have a mismatch. It's awkward to do the accounting when the future resolves because that happens in the base_pipeline code, not write_pipeline code. I'll have to think about it.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77924-->
<!--a209f99e-54e9-418d-a908-743663e6002f-->
<!--title-->
## Retry command for [Build#77924](https://buildkite.com/redpanda/redpanda/builds/77924)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

wdberkeley committed:
ct/l0: Simplify write_and_debounce with reenqueue

write_and_debounce was doing the same logic as push_next_stage /
reenqueue, so replace that logic with the reenqueue method call. Note
that push_next_stage is a method on a stage and there's no stage here to
call it with.

wdberkeley commented:
Force push to rebase on dev and because impl simplified a lot from the original.

wdberkeley committed:
ct/l0: Fix time fallback metric

The time fallback metric was counting requests that were proxied, which
left out requests from the target shard. Plus, with the previous change,
proxied requests may be advanced to batching by the data threshold. This
changes the metric so it's counted when the time fallback mechanism
advances requests to the batcher, which matches how the data threshold
counts requests.
